tuf/repository_lib.py

```diff
@@ -1401,9 +1381,31 @@ def generate_targets_metadata(targets_directory, target_files, version,
     raise securesystemslib.exceptions.Error('Cannot support writing consistent'
         ' targets and using existing fileinfo.')
 
-  if delegations is not None:
+  if delegations:
     tuf.formats.DELEGATIONS_SCHEMA.check_match(delegations)
 
+    # Check if delegated role keys or threshold are changed and update
+    # delegations accordingly. Collect a new delegations key dictionary.
+    keydict = {}
+    delegation_update = False
+    for role in delegations['roles']:
+      rolename = role['name']
+      role_keyids = tuf.roledb.get_role_keyids(rolename, repository_name)
+      role_threshold= tuf.roledb.get_role_threshold(rolename, repository_name)
+
+      if role_keyids != role['keyids'] or \
+          role_threshold != role['threshold']:
+        role['keyids'] = role_keyids
+        role['threshold'] = role_threshold
+        delegation_update = True
+
+      _add_keys_to_keydict(keydict, role_keyids, repository_name)
```

Comment:
Should this be part of the `if` conditional?

Reply:
No, it has to be outside the `if`. The key dictionary has to contain delegations keys of all roles, updated or not.